src/lib/Imp.hs

```diff
@@ -1589,9 +1590,10 @@ impInstrTypes instr = case instr of
 -- TODO: reuse type rules in Type.hs
 impOpType :: IPrimOp n -> IType
 impOpType pop = case pop of
-  ScalarBinOp op x y -> runHardFail $ checkBinOp op (getIType x) (getIType y)
-  ScalarUnOp  op x   -> runHardFail $ checkUnOp  op (getIType x)
-  VectorBinOp op x y -> runHardFail $ checkBinOp op (getIType x) (getIType y)
+  ScalarBinOp op x _ -> typeBinOp op (getIType x)
+  -- All unary ops preserve the type of their input
```

Comment:
While this is true right now, I don't think this is a good idea. Can't we use some type querying function from `QueryType` here? With this we end up with the typing rules for `UnOp`s scattered across 3 files and I'm afraid those will easily get out of sync. You won't get a compiler warning here if you add more cases!

Reply:
Done.